libbeat/outputs/mode/balance_async.go

```diff
@@ -0,0 +1,307 @@
+				if err := client.Connect(m.timeout); err == nil {
+					break
+				}
+
+				if !backoff.Wait() { // done channel closed
+					return
+				}
+			}
+
+			// receive and process messages
+			var msg eventsMessage
+			select {
+			case <-m.done:
+				return
+			case msg = <-m.retries: // receive message from other failed worker
+				debug("events from retries queue")
+			case msg = <-m.work: // receive message from publisher
+				debug("events from worker worker queue")
+			}
+
+			err := m.onMessage(client, msg)
+			if !backoff.WithError(err) { // done channel closed
+				return
+			}
+		}
+	}
+
+	for _, client := range clients {
+		m.wg.Add(1)
+		waitStart.Add(1)
+		go worker(client)
+	}
+	waitStart.Wait()
+}
+
+func (m *AsyncLoadBalancerMode) onMessage(
+	client AsyncProtocolClient,
+	msg eventsMessage,
+) error {
+	var err error
+	if msg.event != nil {
+		err = client.AsyncPublishEvent(handlePublishEventResult(m, msg), msg.event)
+	} else {
+		err = client.AsyncPublishEvents(handlePublishEventsResult(m, msg), msg.events)
+	}
+
+	if err != nil {
+		if msg.attemptsLeft > 0 {
+			msg.attemptsLeft--
+		}
+
+		// asynchronously retry to insert message (if attempts left), so worker can not
+		// deadlock on retries channel if client puts multiple failed outstanding
+		// events into the pipeline
+		m.onFail(true, msg, err)
+	}
+
+	return err
+}
+
+func handlePublishEventResult(m *AsyncLoadBalancerMode, msg eventsMessage) func(error) {
+	return func(err error) {
+		if err != nil {
+			if msg.attemptsLeft > 0 {
+				msg.attemptsLeft--
+			}
+			m.onFail(false, msg, err)
+		} else {
+			outputs.SignalCompleted(msg.signaler)
+		}
+	}
+}
+
+func handlePublishEventsResult(
+	m *AsyncLoadBalancerMode,
+	msg eventsMessage,
+) func([]common.MapStr, error) {
+	total := len(msg.events)
+	return func(events []common.MapStr, err error) {
+		if err != nil {
+			if msg.attemptsLeft > 0 {
+				msg.attemptsLeft--
+			}
+
+			// reset attempt count if subset of messages has been processed
+			if len(events) < total && msg.attemptsLeft >= 0 {
+				msg.attemptsLeft = m.maxAttempts
+			}
+
+			if err != ErrTempBulkFailure {
+				// retry non-published subset of events in batch
+				msg.events = events
+				m.onFail(false, msg, err)
+				return
+			}
+
+			if m.maxAttempts > 0 && msg.attemptsLeft == 0 {
+				// no more attempts left => drop
+				dropping(msg)
+				return
+			}
+
+			// retry non-published subset of events in batch
+			msg.events = events
+			m.onFail(false, msg, err)
+			return
+		}
+
+		// re-insert non-published events into pipeline
+		if len(events) != 0 {
+			msg.events = events
+			if ok := m.forwardEvent(m.retries, msg); !ok {
+				dropping(msg)
+			}
+			return
+		}
+
+		// all events published -> signal success
+		outputs.SignalCompleted(msg.signaler)
+	}
+}
+
+func (m *AsyncLoadBalancerMode) onFail(async bool, msg eventsMessage, err error) {
+	fn := func() {
+		logp.Info("Error publishing events (retrying): %s", err)
```

Comment:
we've got expvar counters on output plugins. As long it's no 'internal drop' due to timeout, error will be counted by output plugins. Some better internal metrics is subject to another PR spanning the other output modes and publishers.